### PR TITLE
Fixing a data correctness issue in unnest when first row of an MVD is null

### PR DESCRIPTION
**Pull request by somu-imply**
This PR solves 2 things:

1. When the first row of a MVD is null, the previous version of unnest was giving an incorrect result as the underlying object was not being updated. 

Previous:
```
select * from mytest1, unnest(mv_to_array(c2)) as unnested(c3)

__time                    c1 c2            c3
2022-01-01T00:00:00.000Z  1  null          null
2022-01-01T00:00:00.000Z  2  ["A","B","C"] null
```

After this change
```
select * from mytest1, unnest(mv_to_array(c2)) as unnested(c3)

__time                    c1 c2            c3
2022-01-01T00:00:00.000Z  1  null          null
2022-01-01T00:00:00.000Z  2  ["A","B","C"] A
2022-01-01T00:00:00.000Z  2  ["A","B","C"] B
2022-01-01T00:00:00.000Z  2  ["A","B","C"] C
```

2. For nested queries that has an unnest in the inner as well as the outer query, the output column name was being always set to EXPR$0. This now changes that behavior to make output column names different.

Before for the query 
```
with t AS (select dim2, d3 from druid.numfoo, unnest(MV_TO_ARRAY(dim3)) as unnested (d3))
select d2,d3 from t, UNNEST(MV_TO_ARRAY(dim2)) as unnested(d2)
```
The planner would do
```
{
  "queryType" : "scan",
  "dataSource" : {
    "type" : "unnest",
    "base" : {
      "type" : "query",
      "query" : {
        "queryType" : "scan",
        "dataSource" : {
          "type" : "unnest",
          "base" : {
            "type" : "table",
            "name" : "numfoo"
          },
          "column" : "dim3",
          "outputName" : "EXPR$0",
          "allowList" : null
        },
        "intervals" : {
          "type" : "intervals",
          "intervals" : [ "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z" ]
        },
        "resultFormat" : "compactedList",
        "columns" : [ "EXPR$0", "__time", "cnt", "d1", "d2", "dim1", "dim2", "dim3", "dim4", "dim5", "dim6", "f1", "f2", "l1", "l2", "m1", "m2", "unique_dim1" ],
        "legacy" : false,
        "context" : {
          "defaultTimeout" : 300000,
          "maxScatterGatherBytes" : 9223372036854775807,
          "sqlCurrentTimestamp" : "2000-01-01T00:00:00Z",
          "sqlQueryId" : "dummy",
          "vectorize" : "false",
          "vectorizeVirtualColumns" : "false"
        },
        "granularity" : {
          "type" : "all"
        }
      }
    },
    "column" : "dim2",
    "outputName" : "EXPR$0",
    "allowList" : null
  },
  "intervals" : {
    "type" : "intervals",
    "intervals" : [ "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z" ]
  },
  "resultFormat" : "compactedList",
  "columns" : [ "EXPR$0", "EXPR$00" ],
  "legacy" : false,
  "context" : {
    "defaultTimeout" : 300000,
    "maxScatterGatherBytes" : 9223372036854775807,
    "sqlCurrentTimestamp" : "2000-01-01T00:00:00Z",
    "sqlQueryId" : "dummy",
    "vectorize" : "false",
    "vectorizeVirtualColumns" : "false"
  },
  "granularity" : {
    "type" : "all"
  }
}
```
After this change the planner plans correctly as 
```
{
  "queryType" : "scan",
  "dataSource" : {
    "type" : "unnest",
    "base" : {
      "type" : "query",
      "query" : {
        "queryType" : "scan",
        "dataSource" : {
          "type" : "unnest",
          "base" : {
            "type" : "table",
            "name" : "numfoo"
          },
          "column" : "dim3",
          "outputName" : "EXPR$0",
          "allowList" : null
        },
        "intervals" : {
          "type" : "intervals",
          "intervals" : [ "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z" ]
        },
        "resultFormat" : "compactedList",
        "columns" : [ "EXPR$0", "__time", "cnt", "d1", "d2", "dim1", "dim2", "dim3", "dim4", "dim5", "dim6", "f1", "f2", "l1", "l2", "m1", "m2", "unique_dim1" ],
        "legacy" : false,
        "context" : {
          "defaultTimeout" : 300000,
          "maxScatterGatherBytes" : 9223372036854775807,
          "sqlCurrentTimestamp" : "2000-01-01T00:00:00Z",
          "sqlQueryId" : "dummy",
          "vectorize" : "false",
          "vectorizeVirtualColumns" : "false"
        },
        "granularity" : {
          "type" : "all"
        }
      }
    },
    "column" : "dim4",
    "outputName" : "EXPR$00",
    "allowList" : null
  },
  "intervals" : {
    "type" : "intervals",
    "intervals" : [ "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z" ]
  },
  "resultFormat" : "compactedList",
  "columns" : [ "EXPR$0", "EXPR$00" ],
  "legacy" : false,
  "context" : {
    "defaultTimeout" : 300000,
    "maxScatterGatherBytes" : 9223372036854775807,
    "sqlCurrentTimestamp" : "2000-01-01T00:00:00Z",
    "sqlQueryId" : "dummy",
    "vectorize" : "false",
    "vectorizeVirtualColumns" : "false"
  },
  "granularity" : {
    "type" : "all"
  }
}
```

Additionally unit test cases for 1 and 2 have been added by creating a new data source in the CalciteTests framework

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by imply-cheddar** [comment]
Did I understand you correctly that the previous (bad?) behavior was

```
select * from mytest1, unnest(mv_to_array(c2)) as unnested(c3)

__time c1 c2 c3
2022-01-01T00:00:00.000Z 1 null null
2022-01-01T00:00:00.000Z 2 ["A","B","C"] A
2022-01-01T00:00:00.000Z 2 ["A","B","C"] B
2022-01-01T00:00:00.000Z 2 ["A","B","C"] C
```

I.e. when it gets to the second row that has an array of 3 values, it unnests it into 3 rows.

And you changed the code to do

```
select * from mytest1, unnest(mv_to_array(c2)) as unnested(c3)

__time c1 c2 c3
2022-01-01T00:00:00.000Z 1 null null
2022-01-01T00:00:00.000Z 2 ["A","B","C"] null
```

I.e. if it sees a null, it will use the null going forward an not unnest anything?

If that understanding is correct, can you explain why the previous behavior is not the correct behavior?  It is what I had expected at least...

**Comment by somu-imply** [comment]
@imply-cheddar it is the other way round the previous was 
```
select * from mytest1, unnest(mv_to_array(c2)) as unnested(c3)

__time c1 c2 c3
2022-01-01T00:00:00.000Z 1 null null
2022-01-01T00:00:00.000Z 2 ["A","B","C"] null
```
Which was incorrect. This was changed to be be the correct one

File: sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java [hunk 1]
```diff
@@ -97,6 +97,7 @@
   public static final String DATASOURCE3 = "numfoo";
   public static final String DATASOURCE4 = "foo4";
   public static final String DATASOURCE5 = "lotsocolumns";
+  public static final String DATASOURCE6 = "unnestnumfoo";
```
Comment (paul-rogers) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java:100]: Better name? `nested` perhaps? Also, would be cool to add a comment with the schema: I find it hard to suss that out from the code.

File: sql/src/test/java/org/apache/druid/sql/calcite/util/TestDataBuilder.java [hunk 2]
```diff
@@ -351,6 +351,100 @@ public Optional<Joinable> build(
   public static final List<InputRow> ROWS1 =
       RAW_ROWS1.stream().map(TestDataBuilder::createRow).collect(Collectors.toList());
 
+  public static final List<ImmutableMap<String, Object>> RAW_ROWS_FOR_UNNEST = ImmutableList.of(
```
Comment (paul-rogers) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/TestDataBuilder.java:354]: Does this have all the interesting corner cases? Empty arrays or objects? Null values? Fields that appear in one nested object but not another (in both orders: (a,b), (a), (a,c))? And so on. To help future readers, might be handy to add a comment above each `.put(` call that sets up one of these cases.

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java [hunk 3]
```diff
@@ -200,11 +200,36 @@ public DruidQuery toDruidQuery(boolean finalizeAggregations)
...
+    //        170:LogicalProject(subset=[rel#171:Subset#2.NONE.[]], EXPR$0=[MV_TO_ARRAY($cor0.dim3)])
+    //          14:LogicalValues(subset=[rel#169:Subset#1.NONE.[0]], tuples=[[{ 0 }]])
+    //  182:Uncollect(subset=[rel#183:Subset#8.NONE.[]])
+    //    180:LogicalProject(subset=[rel#181:Subset#7.NONE.[]], EXPR$0=[MV_TO_ARRAY($cor1.dim2)])
+    //      14:LogicalValues(subset=[rel#169:Subset#1.NONE.[0]], tuples=[[{ 0 }]])
+    //
+    // the column name cannot be EXPR$0 for both inner and outer. The inner one which gets executed first gets the name
+    // EXPR$0 and as we move up the tree we add a 0 at the end to make the top level EXPR$00.
+    // Ideally these names should be replaced by the alias names specified in the query. Any future developer if
+    // able to find these alias names should replace EXPR$0 by dim3 and EXPR$00 by dim2, i.e use the correct name from Calcite
```
Comment (paul-rogers) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java:223]: Thanks much for the detailed explanation!

File: sql/src/test/java/org/apache/druid/sql/calcite/util/TestDataBuilder.java (hunk 2 above)
Comment (somu-imply) [reply sql/src/test/java/org/apache/druid/sql/calcite/util/TestDataBuilder.java:354]: Good idea will do

File: sql/src/test/java/org/apache/druid/sql/calcite/util/TestDataBuilder.java [hunk 4]
```diff
@@ -351,6 +351,100 @@ public Optional<Joinable> build(
...
+      ImmutableMap.<String, Object>builder()
+                  .put("t", "2000-01-01")
+                  .put("m1", "1.0")
+                  .put("m2", "1.0")
+                  .put("d1", 1.0)
+                  .put("f1", 1.0f)
+                  .put("l1", 7L)
+                  .put("dim1", "")
+                  .put("dim3", ImmutableList.of("a", ImmutableList.of("b", "c")))
```
Comment (clintropolis) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/TestDataBuilder.java:363]: the string dimension indexer can't really handle nested arrays like this, i think you'll end up with something like "a" and then the 'toString' of ["b","c"], or maybe something even weirder...

I think you should stick to having either flat lists or single layer strings for these tests

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java [hunk 5]
```diff
@@ -200,11 +200,36 @@ public DruidQuery toDruidQuery(boolean finalizeAggregations)
...
+
+    if (druidQueryRel instanceof DruidCorrelateUnnestRel) {
+      outputColName = outputColName + "0";
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java:226]: im skeptical that this is always correct, is it really cool?

Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java:226]: This is a hacky way as of now, I have kept a pointer to this to be corrected by fetching the actual names. Will do this in a followup PR

**Comment by somu-imply** [comment]
This is fixed through #13934 . Closing in favor of that

**Pull request closed without merging**